src/evm/protocol/lido/state.rs

```diff
@@ -0,0 +1,995 @@
+            }),
+        })
+    }
+
+    fn unwrap_steth(&self, amount_in: BigUint) -> Result<GetAmountOutResult, SimulationError> {
+        if amount_in.is_zero() {
+            return Err(SimulationError::InvalidInput("Cannot unwrap 0 wstETH ".to_string(), None))
+        }
+
+        let amount_out = u256_to_biguint(safe_div_u256(
+            safe_mul_u256(biguint_to_u256(&amount_in), biguint_to_u256(&self.total_pooled_eth))?,
+            biguint_to_u256(&self.total_shares),
+        )?);
+
+        let new_total_wrapped_st_eth = self
+            .total_wrapped_st_eth
+            .as_ref()
+            .expect("total_wrapped_st_eth must be present for wrapped staked ETH pool") -
+            &amount_in;
+
+        Ok(GetAmountOutResult {
+            amount: amount_out.clone(),
+            gas: BigUint::from(DEFAULT_GAS),
+            new_state: Box::new(Self {
+                pool_type: self.pool_type.clone(),
+                total_shares: self.total_shares.clone(),
+                total_pooled_eth: self.total_pooled_eth.clone(),
+                total_wrapped_st_eth: Some(new_total_wrapped_st_eth),
+                id: self.id.clone(),
+                native_address: self.native_address.clone(),
+                stake_limits_state: self.stake_limits_state.clone(),
+                tokens: self.tokens.clone(),
+                token_to_track_total_pooled_eth: self
+                    .token_to_track_total_pooled_eth
+                    .clone(),
+            }),
+        })
+    }
+
+    fn zero2one(&self, sell_token: &Bytes, buy_token: &Bytes) -> Result<bool, SimulationError> {
+        let second_token = self
+            .tokens
+            .iter()
+            .find(|t| **t != self.token_to_track_total_pooled_eth)
+            .expect("No second token found");
+
+        if buy_token == second_token && *sell_token == self.token_to_track_total_pooled_eth {
+            Ok(true)
+        } else if *buy_token == self.token_to_track_total_pooled_eth && sell_token == second_token {
+            Ok(false)
+        } else {
+            Err(SimulationError::InvalidInput(
+                format!(
+                    "Invalid combination of tokens for type {:?}: {:?}, {:?}",
+                    self.pool_type, buy_token, sell_token
+                ),
+                None,
+            ))
+        }
+    }
+
+    fn st_eth_limits(
+        &self,
+        sell_token: Bytes,
+        buy_token: Bytes,
+    ) -> Result<(BigUint, BigUint), SimulationError> {
+        if self.zero2one(&sell_token, &buy_token)? {
+            let limit = self.stake_limits_state.get_limit();
+
+            Ok((limit.clone(), limit))
+        } else {
+            Ok((BigUint::zero(), BigUint::zero()))
+        }
+    }
+
+    fn wst_eth_limits(
+        &self,
+        sell_token: Bytes,
+        buy_token: Bytes,
+    ) -> Result<(BigUint, BigUint), SimulationError> {
+        if !self.zero2one(&sell_token, &buy_token)? {
```

Comment:
Could we add a small comment here indicating which direction this is? like is this wstETH to stETH?